Code/Geometry/Trajectory.cpp

```diff
@@ -0,0 +1,210 @@
+#include <set>
+#include "Trajectory.h"
+
+namespace RDGeom {
+
+Snapshot::Snapshot(double *pos, double energy, bool owner) :
+  d_trajectory(NULL),
+  d_owner(owner),
+  d_energy(energy),
+  d_pos(pos)
+{
+}
+
+Point2D Snapshot::getPoint2D(unsigned int pointNum) const {
+  PRECONDITION(d_pos, "pos must not be NULL");
+  PRECONDITION(trajectory()->dimension() == 2, "d_dimension must be == 2");
+  unsigned int i = pointNum * trajectory()->dimension();
+  return Point2D(d_pos[i], d_pos[i + 1]);
+}
+
+Point3D Snapshot::getPoint3D(unsigned int pointNum) const {
+  PRECONDITION(trajectory()->dimension() >= 2, "d_dimension must be >= 2");
+  unsigned int i = pointNum * trajectory()->dimension();
+  return (d_pos ? Point3D(d_pos[i], d_pos[i + 1],
+          ((trajectory()->dimension() == 3) ? d_pos[i + 2] : 0.0)) : Point3D());
+}
+
+void Snapshot::setTrajectory(const Trajectory *traj) {
+  PRECONDITION(traj, "traj must not be NULL");
+  d_trajectory = traj;
+}
+
+void Snapshot::freePos() {
+  delete d_pos;
+  d_pos = NULL;
+}
+
+Trajectory::Trajectory(unsigned int dimension, unsigned int numPoints) :
+  d_dimension(dimension),
+  d_numPoints(numPoints)
+{
+}
+
+Trajectory::~Trajectory() {
+  for (std::vector<Snapshot>::iterator it = d_snapshotVect.begin();
+    it != d_snapshotVect.end(); ++it) {
+    if (it->getOwner())
+      it->freePos();
+  }
+}
+
+unsigned int Trajectory::addSnapshot(Snapshot s) {
+  unsigned int size = d_snapshotVect.size();
+  s.setTrajectory(this);
+  d_snapshotVect.push_back(s);
+  return size;
+}
+
+const Snapshot &Trajectory::getSnapshot(unsigned int snapshotNum) const {
+  PRECONDITION(snapshotNum < d_snapshotVect.size(), "snapshotNum out of bounds");
+  return d_snapshotVect[snapshotNum];
+}
+
+unsigned int Trajectory::insertSnapshot(unsigned int snapshotNum, Snapshot s) {
+  PRECONDITION(snapshotNum < d_snapshotVect.size(), "snapshotNum out of bounds");
+  s.setTrajectory(this);
+  return (d_snapshotVect.insert(d_snapshotVect.begin() + snapshotNum, s)
+          - d_snapshotVect.begin());
+}
+
+unsigned int Trajectory::removeSnapshot(unsigned int snapshotNum) {
+  PRECONDITION(snapshotNum < d_snapshotVect.size(), "snapshotNum out of bounds");
+  if (d_snapshotVect[snapshotNum].getOwner())
+    d_snapshotVect[snapshotNum].freePos();
+  return (d_snapshotVect.erase(d_snapshotVect.begin() + snapshotNum) - d_snapshotVect.begin());
+}
+
+unsigned int Trajectory::readAmber(const std::string &fName) {
+  PRECONDITION(d_dimension == 3, "The trajectory must have dimension == 3");
+  std::ifstream inStream(fName.c_str());
+  if (!inStream || inStream.bad()) {
+    std::stringstream ss;
+    ss << "Bad input file: " << fName;
+    throw RDKit::BadFileException(ss.str());
+  }
+  std::string tempStr;
+  // title
+  std::getline(inStream, tempStr);
+  // read coordinates
+  unsigned int nCoords = d_numPoints * 3;
+  unsigned int nSnapshots = 0;
+  while (inStream.good() && !inStream.eof()) {
+    double *c = new double[nCoords]();
```

Comment:
I suggest using a shared_ptr for this or potentially a unique_ptr for memory management.

Another method is to have the trajectory hold all the coords and the Snapshots are never owners, just pointers back to the coords.  When the trajectory goes away, the snapshots should be invalidated.